### DISCUSSION TITLE
Making Directus work with Caddy 2

### DISCUSSION THREAD

**Discussion by caleniuc** [directus/directus#2798] [Archive]
OS: Ubuntu 18.04
Web Server: Caddy v2
Directus: Git cloned from [here](https://github.com/directus/directus)

Hello,

I've started creating a new Directus installation and I'm trying to use it with Caddy.
I've managed to install the Database, PHP and I've set up Caddy using this Caddyfile (inspired from [this](https://github.com/caddyserver/examples/blob/master/directus/Caddyfile)):

        sub.domain.com {
            root * /var/www/directus/public
            rewrite /admin /admin/index.html?{query}
            rewrite /thumbnail /thumbnail/index.php?{query}
            @notpath {
			not {
			    path /admin
			    path /storage
			    path /extensions
			    path /thumbnail
			}
            }
            rewrite @notpath /index.php?{query}

            encode gzip zstd
            php_fastcgi unix//run/php/php7.4-fpm.sock
        }

Now when I go to `sub.domain.com` I get redirected to `https://sub.domain.com/admin/` and it's throwing me this error:

`
{"error":{"code":3,"message":"Unauthorized request","class":"Directus\\Exception\\UnauthorizedException","file":"\/var\/www\/directus\/src\/helpers\/app.php","line":287}}
`

I've found [this ](https://github.com/directus/directus/blob/b704bf997f8ce6ff197edbb31b1e933cab2e3bff/src/web.php#L14). I get `401 Unauthorized` on everything except `https://sub.domain.com/server/ping`, which works and returns `pong`.

Is there something more that I maybe missed and I need to configure?
Or probably there's something wrong with my Caddyfile, as v2 changed the syntax and I rewrote it following the example in a bad way.

Thank you :)

**Comment by rijkvanzanten** [comment]
It sounds like it's returning `index.php` even for the `/admin` paths. 

I'm not too familiar with Caddy's ins and outs, but are you sure this:

```
@notpath {
  not {
    path /admin
    path /storage
    path /extensions
    path /thumbnail
  }
}
```

is correct?

**Comment by caleniuc** [comment]
Hi @rijkvanzanten , thanks for your response.

The issue most probably is with the Caddyfile.

I've managed to make it work by using this config:

```
sub.domain.com {
		root * /var/www/directus/public
		file_server

		redir /admin /admin/
		handle /admin/* {
				try_files {path} {path}/ /admin/index.html?{query}
		}

		handle /thumbnail/* {
				try_files {path} {path}/ /thumbnail/index.php?{query}
		}

                handle {
				try_files {path} {path}/ /index.php?{query}
		}

		encode gzip zstd
		php_fastcgi unix//run/php/php7.4-fpm.sock
}

```

I'm not really sure how good it is, but it let me install and now I can login to Directus.

Added file_server

I used [handle ](https://caddyserver.com/docs/caddyfile/directives/handle)because it says in the docs it's similar to location from nginx, and I tried to convert somehow the [nginx config](https://docs.directus.io/advanced/server-setup.html#nginx), but it also worked without `handle` and `try_files` by using `rewrite /admin/* /admin/index.html?{query}`.

I've redirected `/admin` to `/admin/` because otherwise all the files would have been served from `https://sub.domain.com/FILE_PATH_HERE` instead of `https://sub.domain.com/admin/FILE_PATH_HERE`.

And also the path from handle is now `/admin/*` instead of only `/admin`. This I don't understand why it doesn't work with `/admin`

I'll have to read more the caddy docs to understand everything in here and if there is a better way.

**Reply by rijkvanzanten** [reply]
This looks good to me 🙂 

I don't know why there's a difference between `/admin` and `/admin/*`, but the makers of Caddy probably have a good reason for doing that. Glad to hear you figured out a working config 🙂
